### PR TITLE
Generalize Python InputTable class

### PR DESCRIPTION
**Pull request by arman-ddl**
The Python InputTable class is tied to the in-memory Core impls, but it can just accept a table that is supposed to have an input table attribute. This will be beneficial to e.g. Core+ where we can plug in our own implementation and users can use the same experience for both Core and Core+ input tables.

### CODE REVIEW

File: py/server/deephaven/table_factory.py [hunk 1]
```diff
@@ -231,68 +232,48 @@ def write_row(self, *values: Any) -> None:
...
-            self.key_columns = key_cols
+            self.j_input_table.add(table.j_table)
         except Exception as e:
-            raise DHError(e, "failed to create a InputTable.") from e
+            raise DHError(e, "add to InputTable failed.") from e
 
-    def add(self, table: Table) -> None:
-        """Writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
+    def add_async(self, table: Table, j_listener: jpy.JType = _JInputTableStatusListener.DEFAULT) -> None:
+        """Asynchronously  writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:259]: We generally don't expose raw JPY types in public methods. Please refer to the adapter pattern we use for wrapping Python functions/methods for table listeners in `table_listener.py`.

File: py/server/deephaven/table_factory.py [hunk 2]
```diff
@@ -301,18 +282,54 @@ def delete(self, table: Table) -> None:
...
         """
         try:
-            if not self.key_columns:
-                raise PermissionError("deletion on an append-only input table is not allowed.")
             self.j_input_table.delete(table.j_table)
         except Exception as e:
             raise DHError(e, "delete data in the InputTable failed.") from e
 
+    def delete_async(self, table: Table, j_listener: jpy.JType = _JInputTableStatusListener.DEFAULT) -> None:
+        """Asynchronously  deletes the keys contained in the provided table from this keyed input table. If this method is called on an
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:290]: Same comment for `add_async`

File: py/server/deephaven/table_factory.py [hunk 3]
```diff
@@ -301,18 +282,54 @@ def delete(self, table: Table) -> None:
...
+        except Exception as e:
+            raise DHError(e, "delete data in the InputTable failed.") from e
+
+    def get_key_names(self):
+        """Gets the names of the key columns.
+
+        Returns:
+            a list with the names of the key columns of this input table
+        """
+        return self.j_input_table.getKeyNames()
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:311]:
```suggestion
    def get_key_names(self) -> List[str]:
        """Gets the names of the key columns.

        Returns:
            a list with the names of the key columns of this input table
        """
        return j_list_to_list(self.j_input_table.getKeyNames())
```

File: py/server/deephaven/table_factory.py [hunk 4]
```diff
@@ -301,18 +282,54 @@ def delete(self, table: Table) -> None:
...
+
+    def get_value_names(self):
+        """Gets the names of the value columns. By default, any column not marked as a key column is a value column.
+
+        Returns: a list with the names of the value columns of this input table
+        """
+        return self.j_input_table.getValueNames()
+
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:319]:
```suggestion
    def get_value_names(self) -> List[str]:
        """Gets the names of the value columns. By default, any column not marked as a key column is a value column.

        Returns: a list with the names of the value columns of this input table
        """
        return j_list_to_list(self.j_input_table.getValueNames())

```

**Review by jmao-denver — CHANGES_REQUESTED** [review]

**Comment by jmao-denver** [comment]
Please also create some test cases for the new methods in `test_table_factory.py'

File: py/server/deephaven/table_factory.py [hunk 5]
```diff
@@ -301,18 +268,38 @@ def delete(self, table: Table) -> None:
...
-                raise PermissionError("deletion on an append-only input table is not allowed.")
             self.j_input_table.delete(table.j_table)
         except Exception as e:
             raise DHError(e, "delete data in the InputTable failed.") from e
 
+    def get_key_names(self) -> List[str]:
+        """Gets the names of the key columns.
+
+        Returns:
+            a list with the names of the key columns of this input table
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:279]:
```suggestion
    @property
    def get_key_names(self) -> List[str]:
        """The names of the key columns of the InputTable.
```

File: py/server/deephaven/table_factory.py [hunk 6]
```diff
@@ -301,18 +268,38 @@ def delete(self, table: Table) -> None:
...
+        """
+        return j_list_to_list(self.j_input_table.getKeyNames())
+
+    def get_value_names(self) -> List[str]:
+        """Gets the names of the value columns. By default, any column not marked as a key column is a value column.
+
+        Returns: a list with the names of the value columns of this input table
```
Comment (jmao-denver) [inline py/server/deephaven/table_factory.py:286]:
```suggestion
    @property
    def value_names(self) -> List[str]:
        """The names of the value columns. By default, any column not marked as a key column is a value column.

```

File: py/server/tests/test_table_factory.py [hunk 7]
```diff
@@ -327,24 +327,32 @@ def test_input_table(self):
         col_defs = {c.name: c.data_type for c in t.columns}
         with self.subTest("from table definition"):
             append_only_input_table = input_table(col_defs=col_defs)
+            self.assertEqual(append_only_input_table.get_key_names(), [])
+            self.assertEqual(append_only_input_table.get_value_names(), [col.name for col in cols])
```
Comment (jmao-denver) [inline py/server/tests/test_table_factory.py:331]:
```suggestion
            self.assertEqual(append_only_input_table.key_names, [])
            self.assertEqual(append_only_input_table.value_names, [col.name for col in cols])
```

File: py/server/tests/test_table_factory.py [hunk 8]
```diff
@@ -327,24 +327,32 @@ def test_input_table(self):
...
             append_only_input_table.add(t)
             self.assertEqual(append_only_input_table.size, 2)
             append_only_input_table.add(t)
             self.assertEqual(append_only_input_table.size, 4)
 
             keyed_input_table = input_table(col_defs=col_defs, key_cols="String")
+            self.assertEqual(keyed_input_table.get_key_names(), ["String"])
+            self.assertEqual(keyed_input_table.get_value_names(), [col.name for col in cols if col.name != "String"])
```
Comment (jmao-denver) [inline py/server/tests/test_table_factory.py:339]: Same suggestion as earlier.

File: py/server/tests/test_table_factory.py [hunk 9]
```diff
@@ -327,24 +327,32 @@ def test_input_table(self):
...
             keyed_input_table.add(t)
             self.assertEqual(keyed_input_table.size, 2)
             keyed_input_table.add(t)
             self.assertEqual(keyed_input_table.size, 2)
 
         with self.subTest("from init table"):
             append_only_input_table = input_table(init_table=t)
+            self.assertEqual(append_only_input_table.get_key_names(), [])
+            self.assertEqual(append_only_input_table.get_value_names(), [col.name for col in cols])
```
Comment (jmao-denver) [inline py/server/tests/test_table_factory.py:348]: ditto

File: py/server/tests/test_table_factory.py [hunk 10]
```diff
@@ -327,24 +327,32 @@ def test_input_table(self):
...
             self.assertEqual(append_only_input_table.size, 2)
             append_only_input_table.add(t)
             self.assertEqual(append_only_input_table.size, 4)
 
             keyed_input_table = input_table(init_table=t, key_cols="String")
+            self.assertEqual(keyed_input_table.get_key_names(), ["String"])
+            self.assertEqual(keyed_input_table.get_value_names(), [col.name for col in cols if col.name != "String"])
```
Comment (jmao-denver) [inline py/server/tests/test_table_factory.py:355]: ditto

File: py/server/tests/test_table_factory.py [hunk 11]
```diff
@@ -355,9 +363,11 @@ def test_input_table(self):
             append_only_input_table = input_table(init_table=t)
             with self.assertRaises(DHError) as cm:
                 append_only_input_table.delete(t)
-            self.assertIn("not allowed.", str(cm.exception))
+            self.assertIn("doesn\'t support delete operation", str(cm.exception))
 
             keyed_input_table = input_table(init_table=t, key_cols=["String", "Double"])
+            self.assertEqual(keyed_input_table.get_key_names(), ["String", "Double"])
+            self.assertEqual(keyed_input_table.get_value_names(), [col.name for col in cols if col.name != "String" and col.name != "Double"])
```
Comment (jmao-denver) [inline py/server/tests/test_table_factory.py:370]: ditto

**Review by jmao-denver — CHANGES_REQUESTED** [review]

File: py/server/deephaven/table_factory.py (hunk 5 above)
Comment (arman-ddl) [reply py/server/deephaven/table_factory.py]: I'm not very familiar w/ Python `@property`, but it looks like this method should now be called `key_names`, not `get_key_names`. Changing.

File: py/server/deephaven/table_factory.py [hunk 12]
```diff
@@ -33,6 +32,7 @@
 _JRingTableTools = jpy.get_type("io.deephaven.engine.table.impl.sources.ring.RingTableTools")
 _JSupplier = jpy.get_type('java.util.function.Supplier')
 _JFunctionGeneratedTableFactory = jpy.get_type("io.deephaven.engine.table.impl.util.FunctionGeneratedTableFactory")
+_JInputTableStatusListener = jpy.get_type("io.deephaven.engine.util.input.InputTableStatusListener")
```
Comment (rcaudy) [inline py/server/deephaven/table_factory.py:35]: I think this is unused now.

Comment (chipkent) [reply py/server/deephaven/table_factory.py:35]: I agree.

File: py/server/deephaven/table_factory.py [hunk 13]
```diff
@@ -301,18 +268,36 @@ def delete(self, table: Table) -> None:
...
-            if not self.key_columns:
-                raise PermissionError("deletion on an append-only input table is not allowed.")
             self.j_input_table.delete(table.j_table)
         except Exception as e:
             raise DHError(e, "delete data in the InputTable failed.") from e
 
+    @property
+    def key_names(self) -> List[str]:
+        """The names of the key columns of the InputTable.
+        """
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:278]:
```suggestion
        """The names of the key columns of the InputTable."""
```

File: py/server/deephaven/table_factory.py [hunk 14]
```diff
@@ -301,18 +268,36 @@ def delete(self, table: Table) -> None:
...
+        """The names of the value columns. By default, any column not marked as a key column is a value column.
+
+        """
+        return j_list_to_list(self.j_input_table.getValueNames())
+
 
 def input_table(col_defs: Dict[str, DType] = None, init_table: Table = None,
                 key_cols: Union[str, Sequence[str]] = None) -> InputTable:
-    """Creates an InputTable from either column definitions or initial table. When key columns are
+    """Creates an in-memory InputTable from either column definitions or initial table. When key columns are
```
Comment (chipkent) [inline py/server/deephaven/table_factory.py:291]:
```suggestion
    """Creates an in-memory InputTable from either column definitions or an initial table. When key columns are
```

**Review by chipkent — APPROVED** [review]

**Review by jmao-denver — APPROVED** [review]
LGTM

**Pull request merged**